torch/csrc/jit/codegen/cuda/scheduler.cpp

```diff
@@ -670,6 +874,368 @@ void scheduleReduction(
+        }
+      }
+      if (!constant_allocation) {
+        computeAt_inline_tv.push_back(tv);
+      }
+    }
+  }
+  return computeAt_inline_tv;
+}
+
+//! Place all cache TensorViews in Shared Memory
+//! All point-wise TensorViews inherit shared memory from their parents
+void setupSharedMemory(
+    Fusion* fusion,
+    const std::vector<TensorView*>& cache_tv) {
+  std::vector<TensorView*> stack(cache_tv.begin(), cache_tv.end());
+  while (!stack.empty()) {
+    auto tensor = stack.back();
+    stack.pop_back();
+    if (!fusion->hasOutput(tensor) && !fusion->hasInput(tensor)) {
+      tensor->setMemoryType(MemoryType::Shared);
+      for (auto expr : fusion->unordered_uses(tensor)) {
+        if (isPointwiseOp(expr)) {
+          auto output = expr->output(0)->as<TensorView>();
+          stack.push_back(output);
+        }
+      }
+    }
+  }
+}
+
+} // namespace
+
+void scheduleMultipleReduction(
+    Fusion* fusion,
+    const ReductionParams& rparams,
+    const std::vector<TensorView*>& reduction_tv,
+    std::vector<TensorView*>& other_tv) {
+  FusionGuard fg(fusion);
+
+  const auto& inputs_tv = ir_utils::filterByType<TensorView>(fusion->inputs());
+  const auto& outputs_tv =
+      ir_utils::filterByType<TensorView>(fusion->outputs());
+
+  // Scheduling the Reduction
+  if (rparams.fastest_dim) {
+    if (rparams.multiple_reds_per_blk) {
+      // 1) If Persistent Kernel: Automatically cache input in shared memory
+      // Else: Find TensorViews to duplicate and computeAt inline
+      if (!rparams.persistent_kernel) {
+        auto duplicate_tv = findTensorViewsToDuplicate(fusion, other_tv);
+
+        // Any TVs with multiple uses and dependencies with same IterDomain
+        // Order of Duplication is necessary for correctness
+        for (auto tensor : duplicate_tv) {
+          auto result = tensor->duplicate();
+          other_tv.insert(other_tv.end(), result.begin(), result.end());
+        }
+      }
+
+      // 2) Apply heuristics to each reduction
+      std::vector<TensorView*> rfactor_tv;
+      for (auto tv : reduction_tv) {
+        // Reduction Splits
+        // [output-Leftover, Batches-Per-Block, Warp, Loop-Unroll]
+        if (rparams.batches_per_block > 1) {
+          tv->split(0, rparams.batches_per_block);
+        }
+        if (rparams.num_warps > 1) {
+          tv->split(1, rparams.num_warps);
+        }
+        tv->split(-1, rparams.loop_unroll);
+
+        auto reduction_tv_rf = tv->rFactor({-1});
+        rfactor_tv.push_back(reduction_tv_rf);
+      }
+
+      // 3) Split the other TensorViews
+      for (auto tv : other_tv) {
+        if (rparams.batches_per_block > 1) {
+          tv->split(0, rparams.batches_per_block);
+        }
+        if (rparams.num_warps > 1) {
+          tv->split(1, rparams.num_warps);
+        }
+        tv->split(-1, rparams.loop_unroll);
+      }
+
+      // 4) ComputeAt Structure
+      const int kComputeAtAxis = 1;
```

Comment:
I think this can be 2 if `rparams.batches_per_block > 1` there was an smem bug before that probably prevented this from working, but should be fixed now.